cpp/template/src/ivf_flat_example.cu

```diff
@@ -0,0 +1,173 @@
+
+  int seed = 137;
+  raft::random::RngState rng(seed);
+  auto data_indices  = raft::make_device_vector<int64_t>(dev_resources, n_samples);
+  auto train_indices = raft::make_device_vector<int64_t>(dev_resources, n_train);
+
+  thrust::counting_iterator<int64_t> first(0);
+  thrust::device_ptr<int64_t> ptr(data_indices.data_handle());
+  thrust::copy(raft::resource::get_thrust_policy(dev_resources), first, first + n_samples, ptr);
+
+  raft::random::sample_without_replacement(dev_resources,
+                                           rng,
+                                           raft::make_const_mdspan(data_indices.view()),
+                                           std::nullopt,
+                                           train_indices.view(),
+                                           std::nullopt);
+
+  raft::matrix::copy_rows(
+    dev_resources, dataset, trainset.view(), raft::make_const_mdspan(train_indices.view()));
+
+  return trainset;
+}
+
+void ivf_flat_build_extend_search(raft::device_resources const& dev_resources,
+                                  raft::device_matrix_view<const float, int64_t> dataset,
+                                  raft::device_matrix_view<const float, int64_t> queries)
+{
+  using namespace raft::neighbors;
+
+  // Sub-sample the dataset to create a training set.
+  auto trainset = subsample(dev_resources, dataset, 0.1);
+
+  ivf_flat::index_params index_params;
+  index_params.n_lists           = 100;
+  index_params.metric            = raft::distance::DistanceType::L2Expanded;
+  index_params.add_data_on_build = false;
+
+  std::cout << "\nRun k-means clustering using the training set" << std::endl;
+  auto index =
+    ivf_flat::build(dev_resources, index_params, raft::make_const_mdspan(trainset.view()));
+
+  std::cout << "Number of clusters " << index.n_lists() << ", number of vectors added to index "
+            << index.size() << std::endl;
+
+  std::cout << "Filling index with the dataset vectors" << std::endl;
+
+  auto data_indices = raft::make_device_vector<int64_t, int64_t>(dev_resources, dataset.extent(1));
+  index             = ivf_flat::extend(dev_resources,
+                           dataset,
+                           std::make_optional(raft::make_const_mdspan(data_indices.view())),
+                           index);
+
+  std::cout << "Index size after addin dataset vectors " << index.size() << std::endl;
+
+  // Set search parameters.
+  ivf_flat::search_params search_params;
+  search_params.n_probes = 10;
+
+  // Create output arrays.
+  int64_t topk      = 10;
+  int64_t n_queries = queries.extent(0);
+  auto neighbors    = raft::make_device_matrix<int64_t, int64_t>(dev_resources, n_queries, topk);
+  auto distances    = raft::make_device_matrix<float, int64_t>(dev_resources, n_queries, topk);
+
+  // Search K nearest neighbors for each queries.
+  ivf_flat::search(
+    dev_resources, search_params, index, queries, neighbors.view(), distances.view());
+}
```

Comment:
Would it make sense to return the `make_tuple(neighbors, distances)` here or just print to `std::cout` a few neighbors?
This would also be educational to the viewer in that we'd have to call `dev_resources.sync_stream()` before accessing the results from the host.

Reply:
Good point about highlighting `sync_stream`! Added together with a print function that list neighbors of first queries.